designs/nodeclaim-limits.md

```diff
@@ -0,0 +1,54 @@
+
+This works for users who are purely concerned about limiting resource usage within a NodePool, but doesn't work for specifying a global limit across all NodePools or for limiting a higher level representation of usage such as the number of nodes in a cluster (without also reducing flexibility of instance types).
+
+## Use-Cases
+1. I'm using a shared network with limited IP pools and need to ensure Karpenter's usage doesn't consume too many IPs. Each Node consumes a minimum number of IPs but low-level resource usage for things like CPU/Memory differs between instance types.
+2. I have a special NodePool (for example a set of high performance nodes or AWS Reserved Instances) with a finite amount of compute available. I need a way to prevent Karpenter from launching too many nodes of this type. The only to do this is by calculating against an avialable resource like number of CPU's and adding this as a limit. This doesn't target the actual limiter which is number of instances rather than resource consumption.
+3. My CNI can't support more than 400 nodes but I'm not concerned about resource usage. I have no way to restrict Karpenter to this limit without severly limiting instance type options.
+
+
+Currently, Karpenter users face a limitation in controlling the maximum number of nodes provisioned by Karpenter. This constraint hinders operations in environments where the CNI in use can't scale beyond a certain number of nodes or in scenarios involving shared networks with a finite IP address allocation. Some additional use cases and discussion on this limitation can be seen in the following issues: (#732, aws/karpenter-provider-aws#4462).
+
+## Solutions
+
+### [Recommended] Solution 1: New controller flag `max-node-claims` and per-NodePool limits in the `spec.limits` section of NodePool template
+
+1. A new controller startup flag `max-node-claims` which would allow for setting a global cap on the number of nodes provisioned via this Karpenter controller.
+2. Allow for setting per-NodePool limits on number of NodeClaims. 
+
+```
+--max-node-claims="-1" Usage: The maximum allowed number of NodeClaims that if exceeded will stop additional creations. Negative values are treated as unlimited and is the default.
+```
+
+```
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+spec:
+  template:
+    spec:
+        limits: 
+            nodes: 10
+```
+
+#### Pros/Cons
+
+1. 👍👍 Allows for simple cap on number on NodeClaims created by Karpenter, allowing users to apply this globally and per-NodePool
+2. 👎 May be used as a crutch vs having the cloudprovider providing the relevant resources. An example here could be supporting a limit for IP usage.
+
+### Solution 2: Suggest users request/implement reporting of resource types in CloudProvider
+
+In this solution, the community would need to update Cloud Providers to collect and report on resource usage of the desired types.
+
+#### Pros/Cons
```

Comment:
could you provide more details with this approach? I don't fully understand it.

Reply:
I'll rework this a bit as there's an argument to push for both independently.
As a summary though, for certain use cases such as wanting to limit IP usage it may be possible to instead expose this information available under node resources by working with CNIs. I'm not sure how feasible this specific example is but the point I'm trying to make is that while limiting node count resolves the use case, it may not be addressing support for the actually desired resource to limit on (e.g. IPs).

Still, there are valid use cases for purely limiting node count such as your mention in the PR about Kubenet's limitation on scaling beyond 400 nodes. I think that Karpenter should support setting node count limits but users should be encouraged to open feature requests against the relevant projects they'd like to have more resource information from.

Reply:
I've removed the second proposal but can add additional options if you'd like. I think no matter what there's still valid use cases for having a configurable limit on number of nodes so don't want to distract discussions.